### PR TITLE
Add dataset config file attribute discovery files

### PR DESCRIPTION
**Pull request by daniel-ji**
Addresses https://github.com/chanzuckerberg/cryoet-data-portal/issues/825

Created a python script that merges all dataset configuration files into one large, attribute-comprehensive yaml file for use when generating / updating the template.yaml file. Created as part of the push for dataset configuration file validation. Let me know what changes need to be made! Thank you!

### CODE REVIEW

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        if file.endswith(".yaml") and file not in EXCLUDE_LIST:
+            with open(file, "r") as stream:
+                try:
+                    config_file = yaml.safe_load(stream)
+                except yaml.YAMLError as exc:
+                    print(exc)
+                except Exception:
+                    print("Error in file: ", file)
+                finally:
+                    stream.close()
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:196]: This is a really good practice. 👏  
One nice thing with Python's context manager is that it closes the file, when it exits the block. 

Here is some more information on it: https://medium.com/@orichcaspertheghost/pythons-with-statement-beyond-just-opening-and-closing-files-f9edf368a3be

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+                unified_config = recursive_dict_update(unified_config, config_file)
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:198]: A minor optimization would be to move this to inside the try block or add a continue to the exception blocks, as we will be running this against a config file that has already been processed if we encounter an error or exception.

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+"""
+This function is the main function that will be called. It will take in the current dictionary and the new dictionary, and will merge them together. A recursive
+approach is used to handle nested dictionaries and lists. The function will return the merged dictionary.
+"""
+
+
+def recursive_dict_update(current_entries, new_entries):
+    for key, new_value in new_entries.items():
+        # Regular scenarios
+        if new_value is None:
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:126]: We should still include fields that have the value set as None. As they could be an expected key in the ingestion processing.

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            continue
+        # Dict and dict, so recurse
+        if isinstance(new_value, dict) and (
+            isinstance(current_entries.get(key), dict) or current_entries.get(key) is None
+        ):
+            current_entries[key] = recursive_dict_update(current_entries.get(key, {}), new_value)
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:132]: Nicely done! :)

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                new_list = [new_value] + current_entries.get(key)
+                # and then now it is list, list situation
+                current_entries = recursive_dict_update_list_helper(current_entries, key, new_list)
+            # edge case: one-length list and a non-list, non-dict new_value = just keep the list
+            elif (
+                len(current_entries.get(key)) == 1
+                and not isinstance(new_value, list)
+                and not isinstance(new_value, dict)
+            ):
+                pass
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:163]: Can we add a print here as well?

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            else:
+                print("list conflict")
+                print("Key: ", key)
+                print("Current: ", current_entries.get(key))
+                print("New: ", new_value)
+        else:
+            raw_data_check(current_entries.get(key, None), new_value)
+            current_entries[key] = new_value
```
Comment (manasaV3) [inline ingestion_tools/dataset_configs/config-attribute-discovery.py:171]: What are your thoughts on capturing the types of values passed for this field?

**Review by manasaV3 — COMMENTED** [review]
Great work!

File: ingestion_tools/dataset_configs/config-attribute-discovery.py (hunk 3 above)
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:126]: Noted! linkml may treat a value set as None as not existing (like the attribute was never provided in the config file), but I'll have to look into this more.

File: ingestion_tools/dataset_configs/config-attribute-discovery.py (hunk 5 above)
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:163]: I realized this edge case wasn't written correctly, fixed now and there will be warnings logged if type conflicts occur!

File: ingestion_tools/dataset_configs/config-attribute-discovery.py (hunk 6 above)
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:171]: Done!

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        os.path.join(directory_path, file)
+        for directory_path, _, filename in os.walk(os.path.expanduser("./"))
+        for file in filename
+    ]
+
+    unified_config = {}
+
+    # get all yaml files, except for the whitelist, and attempt to merge them
+    for file in all_files:
+        if file.endswith(".yaml") and file not in EXCLUDE_LIST:
```
[the remark this answers is not in the document]
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:187]: Done!

File: ingestion_tools/dataset_configs/config-attribute-discovery.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
+# TODO: add which files each attribute appears in? and what % of files it appears in?
+"""
+This file is intended to be run from the ingestion/dataset_configs directory. Intended for ensuring that all existing dataset config files' schema are covered by the linkml/Pydantic models.
```
[the remark this answers is not in the document]
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:3]: Done!

File: ingestion_tools/dataset_configs/config-attribute-discovery.py (hunk 2 above)
Comment (daniel-ji) [reply ingestion_tools/dataset_configs/config-attribute-discovery.py:198]: Done!

**Comment by daniel-ji** [comment]
This PR is outdated, I think it may be better to just incrementally PR changes that are currently getting implemented in this branch: https://github.com/daniel-ji/cryoet-data-portal-backend/tree/dataset-config-validation-workflow

**Pull request closed without merging**